### ISSUE TITLE
Azurerm_frontdoor with v2.24.0 breaks when azure frontdoor is edited in portal. 

### ISSUE THREAD

**Issue by andrstor** [terraform-providers/terraform-provider-azurerm#8208]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.12.21
+ provider.azurerm v2.24.0
```

### Affected Resource(s)

* `azurerm_frontdoor`

### Terraform Configuration Files

```hcl
provider "azurerm" {
  version = "=2.24.0"
  features {} # https://www.terraform.io/docs/providers/azurerm/index.html#features
}

resource "azurerm_resource_group" "example" {
  name     = "andreastester"
  location = "norway east"
}

resource "azurerm_frontdoor" "example" {
  name                                         = "andreastester"
  resource_group_name                          = azurerm_resource_group.example.name
  enforce_backend_pools_certificate_name_check = false

  routing_rule {
    name               = "exampleRoutingRule1"
    accepted_protocols = ["Http", "Https"]
    patterns_to_match  = ["/*"]
    frontend_endpoints = ["exampleFrontendEndpoint1"]
    forwarding_configuration {
      forwarding_protocol = "MatchRequest"
      backend_pool_name   = "exampleBackendBing"
    }
  }

  backend_pool_load_balancing {
    name = "exampleLoadBalancingSettings1"
  }

  backend_pool_health_probe {
    name = "exampleHealthProbeSetting1"
    
  }

  backend_pool {
    name = "exampleBackendBing"
    backend {
      host_header = "www.bing.com"
      address     = "www.bing.com"
      http_port   = 80
      https_port  = 443
    }

    load_balancing_name = "exampleLoadBalancingSettings1"
    health_probe_name   = "exampleHealthProbeSetting1"
  }

  frontend_endpoint {
    name                              = "exampleFrontendEndpoint1"
    host_name                         = "andreastester.azurefd.net"
    custom_https_provisioning_enabled = false
  }
}
```

### Debug Output

https://gist.github.com/andrstor/0aa07440e0a01befb23351db3257340f

### Expected Behavior

Terraform identifies that no changes are required or tries to recover its state.

### Actual Behavior

Error: flattening `backend_pool`: ID was missing the `healthProbeSettings` element

### Steps to Reproduce

1. `terraform apply`
2. Do anything in the azure portal that trigges a change. For instance add a rule engine rule to the routing rule. 
3. `terraform plan`

You can also undo the manual change again, the resource is still broken for azurerm v2.24.0. This works with v2.23.0

### Important Factoids

None

### References

* #8046

**Comment by eliasgrueninger** [comment]
Same issue here but on another level:
Error: flattening `frontend_endpoint`: ID was missing the `frontDoorWebApplicationFirewallPolicies` element

**Comment by lyubomirr** [comment]
Same behaviour when trying to import the resource.

**Comment by MelHarbour** [comment]
Seeing the same issue as @eliasgrueninger on the Firewall Policies element, but we're running 2.21.

**Comment by nmiodice** [comment]
I found a similar issue (#8231) and did not realize there was an open issue already.

The only difference in my case is that the AFD & related resources were not edited in the Azure Portal - these resources are fully managed by Terraform. cc @tombuildsstuff , that is the only difference I see between these two issues, though the root cause is likely the same.

**Comment by kevinchabreck** [comment]
I experienced this issue when trying to update the resource, and then again when trying to import it after removing it from the terraform state.

```
import azurerm_frontdoor.fd <my frontdoor resource ID>
azurerm_frontdoor.fd: Importing from ID "<my frontdoor resource ID>"...
azurerm_frontdoor.fd: Import prepared!
  Prepared azurerm_frontdoor for import
azurerm_frontdoor.fd: Refreshing state... [id=<my frontdoor resource ID>]

Error: flattening `backend_pool`: ID was missing the `healthProbeSettings` element
```

I tested this on terraform `v0.12.24` with azurem provider versions `v2.13.0` and `v2.24.0` with the same results. Perhaps @MelHarbour is right about it being a change in Azure's API?

**Comment by lyubomirr** [comment]
@kevinchabreck I've had the exact same issue with the same steps as you, but it started working once i changed the provider version to 2.23.0. You can try it.

**Comment by MelHarbour** [comment]
As a follow-on, we re-ran a previous release on an environment that hadn't been modified outside of Terraform since the last apply, and it also failed with the same error. So it looks distinctly like something's changed in the underlying platform.

**Comment by kevinchabreck** [comment]
@lyubomirr downgrading to `azurerm v2.23.0` seemed to work! Additionally, I had previously modified my resource ID string during the import from `.../providers/Microsoft.Network/frontdoors/...` to `.../providers/Microsoft.Network/frontDoors/...` due to an error I got when attempting a previous import. Reverting this and importing the resource ID exactly as it is shown in the Azure console **(ie. the one with the lowercase D in `frontdoors`)** fixed this issue. Thanks for the tip!

**Comment by MelHarbour** [comment]
Did another bit of testing - realised that we _were_ actually using 2.24, so I've also downgraded to 2.23 and it appears to be working again, so looks like the regression is in provider version 2.24.

**Comment by MelHarbour** [comment]
Presumably the issue was introduced in https://github.com/terraform-providers/terraform-provider-azurerm/pull/8146 as part of the rewriting around IDs.

**Comment by tombuildsstuff** [comment]
@MelHarbour as you mentioned above, unfortunately the FrontDoor API is broken here in that it's returning these in the incorrect case.

Azure API's are supposed to be case-insensitive for Requests but that URI's listed in Responses should be case-sensitive - the Resource Group can be case-insensitive, but unfortunately the way this has been implemented means that the entire URI can be insensitive at Request time.

Other downstream API's treat URI's in a mixed manner - where some API's require that URI's are case sensitive - and others aren't bothered by casing - as such we intentionally treat them as case-sensitive (since that's how the HTTP spec recommends treating URI's).

Unfortunately the FrontDoor API implements this in such a fashion that they're case-insensitive in the response too (specifically being lower-cased when updated through the Portal) - such that #8146 tries to make this consistent by updating/requiring the ID's to be in a consistent format (since users should be able to rely on these being the same). As such it appears there's some more edge-cases to handle here, as a follow up to #8146 - would anyone be able to provide an example of the ID formats being returned/in the state for these fields?

Thanks!

**Comment by andrstor** [comment]
@tombuildsstuff I dont know if this is what you are asking, but I have noticed this behaviour:

After creating the frontdoor resource with azurerm v2.24 it looks like [this](https://gist.github.com/andrstor/92b2b67618c709a3de790f65215a6f41) in https://resources.azure.com/.

After editing it manually in the portal (whatever change), it looks like [this](https://gist.github.com/andrstor/20547389cf52292a2dab37c1f0d77db0). Notice how many of the ID's have all become lowercase.

When running terraform plan again now, it fails. You can actually edit the resource in https://resources.azure.com/ manually (edit the template and use PUT), and terraform will start working again (if you managed to correctly adjust all the lowercase ID's.

**Comment by tombuildsstuff** [comment]
@andrstor perfect, thanks - where the entire ID is being lower-cased, that's a bug in the API (since the health probe names etc should be consistent) - so we'll need to raise an API bug here either way unfortunately

**Comment by robselway** [comment]
Hi @tombuildsstuff - just to clarify - is the only resolution here to wait for Azure to fix the issue? I'm trying to figure out whether it's worth replacing this resource with another script until it's resolved.

**Comment by cpressland** [comment]
We fought so hard with Azure Support during some previous Azure Front Door Terraform/API issues to get them to recognise the Azure API was a bit of a mess and provided multiple examples via Terraform, Azure Portal, and Azure CLI. Response was simply that this isn't an issue because the Azure Portal still works, I kinda get it, but I equally don't think it's this projects responsibility to have to constantly build work-arounds to a buggy API. I'll raise this issue with our Account Managers etc again and see if we can get any traction.

**Comment by tombuildsstuff** [comment]
@robselway 

> just to clarify - is the only resolution here to wait for Azure to fix the issue?

Based on what I can see, unfortunately yes.

The [Azure API Specification](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#request) states that values should be returned in the casing they're submitted (although the HTTP Specification states URI's should be case-sensitive but I digress) - so unfortunately this is an API bug which needs to be fixed, since this should be returned in the same casing we're submitting it in here.

For what it's worth we've also raised this on our end - unfortunately the Networking API's differ from every other Azure API here, so I don't think we can easily work around this (unless perhaps we can find the original casing from the specific sub-element, but that's assuming the API doesn't change to break the casing there too)

---

@JeffreyRichter this is a good example of the Networking API's returning URI's in a case-insensitive manner which [differs from the recommendation in the ARM Spec](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#request) (which unfortunately differs from the HTTP Specification, where the entire URI is case sensitive, [this StackOverflow answer for more details](https://stackoverflow.com/a/26196170)).

Whilst it'd be possible to work around this bug if the "resource type" segment could be parsed case-insensitively, where the entire URI is lower-cased in some responses, but not in others ([see this comment for example responses](https://github.com/terraform-providers/terraform-provider-azurerm/issues/8208#issuecomment-681716403)) - there's not much we can do here, since we can't guarantee these are the correct casing (or that the Networking API's won't change this casing in a future update).

Whilst this Github issue is the wrong place for this discussion, I feel like perhaps the current ambiguous behaviour defined in the ARM Specification ("return in the casing the user passed it in as") is the cause of this confusion - perhaps it'd be clearer if the ARM Specification stated that the entire URI must (to use the language of RFC 7230) be treated as case-sensitive in Responses - WDYT?

Thanks!

**Comment by SunnyOswal** [comment]
Facing same issue with azurerm Provider: 2.26.0 .

**Comment by tombuildsstuff** [comment]
@WodansSon would you be able to get an ETA for a fix from the service team here?

**Comment by WodansSon** [comment]
> @WodansSon would you be able to get an ETA for a fix from the service team here?

@tombuildsstuff I will reach out to the Front Door service team and see how quickly we can get a fix in place for this issue, it will most likely include some cross team collaboration with the portal team to roll back their changes to get this issue totally fixed.

**Comment by nmiodice** [comment]
Is there a viable workaround that we can use until its fixed?

**Comment by surlypants** [comment]
> Is there a viable workaround on the TF side that we can use until it's fixed?

I could not simply just downgrade. I had to import the portal-modified resource into a temporary workspace then push that state file to the appropriate place and _then_ downgrade.

**Comment by nmiodice** [comment]
@surlypants can you elaborate on this?
>push that state file to the appropriate place

**Comment by surlypants** [comment]
> @surlypants can you elaborate on this?
> 
> > push that state file to the appropriate place

I'll try...

The (remote) state file in the terragrunt workspace containing our FD could not plan (with 2.25.0) post portal touch. Downgrading to 2.23.0 still would not pass a plan. I thus imported the resource to a local state file, downloaded the remote state file and replaced its front door module's "instances" array with that from the local state file. I then state push-ed the result back up to the remote backend. Finally, 2.23.0 would plan. We now are back to the time-outs workaround from:

https://github.com/terraform-providers/terraform-provider-azurerm/issues/7925

Side note:
I tried setting required_providers:
azurerm = ">= 2.23.0, <=2.25.0"
at the top level workspace and scoped 2.23.0 specifically to the frontdoor workspace; but it seems that when providing a range, the lowest always wins. So our entire infra is back to 2.23.0

hope this helps / is understandable

**Comment by nmiodice** [comment]
Definitely, thank you @surlypants !

**Comment by phenggeler** [comment]
Hello, is there an update on ETA for this issue? We are not able to manage Frontdoor via Terraform while this issue persists.

**Comment by naikajah** [comment]
we are also hitting this issue when migrating to 2.24. Could we please get an update as to when the bug will be fixed?

**Comment by jmcshane** [comment]
Is there anything that can be done once you face this error with a provider greater than 2.24? I'm getting:

```
Error: Resource instance managed by newer provider version

The current state of module.front_door.azurerm_frontdoor.frontdoor was created
by a newer provider version than is currently selected. Upgrade the azurerm
provider to work with this state.


Error: Resource instance managed by newer provider version

The current state of
module.front_door.azurerm_frontdoor_firewall_policy.policy was created by a
newer provider version than is currently selected. Upgrade the azurerm
provider to work with this state.
```

Can you delete the backend settings and reset the front door? Can you delete the front door and have it recreated?

**Comment by GarethOates** [comment]
I'm facing the same problem as @jmcshane where I cannot use a provider newer than 2.23.0.  This is causing me problems when I want to use features released in later versions, like 2.27.0 for example.  I want to know that it's safe to recreate the resource using a newer provider.

**Comment by Lahiri** [comment]
I'm also struggling with this problem, is a fix in the works?

**Comment by scottzilla** [comment]
@GarethOates @Lahiri you should be able to bring in a provider of a different version using provider aliases, but you'll need to explicitly reference them where appropriate.

See: https://www.terraform.io/docs/configuration/providers.html#alias-multiple-provider-configurations

**Comment by GarethOates** [comment]
> @GarethOates @Lahiri you should be able to bring in a provider of a different version using provider aliases, but you'll need to explicitly reference them where appropriate.
> 
> 
> 
> See: https://www.terraform.io/docs/configuration/providers.html#alias-multiple-provider-configurations

Yeah I tried that but I think as someone mentioned before it just reverts to the oldest version. It didn't work for me anyway even though on unit it downloaded both provider versions.

**Comment by CoopCNIT** [comment]
I tried to run this provider setup to come around it:

```
provider "azurerm" { 
  version         = "2.15.0"
  features {}
  subscription_id = var.subscription_id
  client_id       = var.client_id
  client_secret   = var.client_secret
  tenant_id       = var.tenant_id
}
provider "azurerm" { 
  alias           = "temp"
  version         = "2.30.0"
  features {}
  subscription_id = var.subscription_id
  client_id       = var.client_id
  client_secret   = var.client_secret
  tenant_id       = var.tenant_id
}
```

..but it fails with this error: 

```
Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/azurerm: no available releases match the given constraints 2.15.0,
2.30.0
```

**Comment by cpressland** [comment]
@scottzilla We already have a ton of subscription and use Terraform provider Aliases extensively. Unfortunately it doesn't let us mix and match Provider versions though, I get the same error as @CoopCNIT.

**Comment by IanMoroney** [comment]
@WodansSon , Is there any update from the Azure Front Door service team?

**Comment by scottzilla** [comment]
Sorry to mislead, I could've sworn I had done this in the past.  You are correct, it does appear to resolve a provider with the superset of the version spec, yuck.

**Comment by cpressland** [comment]
I decided to take some time to try and fix this tonight, after further investigation Case Sensitivity is the only issue in play here. A fix was put in for this in with #8046 but this was ultimately rejected. I started by upgrading to the latest release and removing my Front Door from the Terraform State and attempting to reimport.

Unfortunately this failed, but with something new:
```
➜  terraform import "module.uksouth-frontdoor.azurerm_frontdoor.frontdoor" "/subscriptions/<snipped>/resourceGroups/frontdoor/providers/Microsoft.Network/frontdoors/bink-frontdoor"

Error: Error parsing Resource ID "/subscriptions/<snipped>/resourceGroups/frontdoor/providers/Microsoft.Network/frontdoors/bink-frontdoor": ID was missing the `frontDoors` element
```

Changing to the case Terraform is expecting returns us back to a broken, but familiar state.
```
➜  terraform import "module.uksouth-frontdoor.azurerm_frontdoor.frontdoor" "/subscriptions/<snipped>/resourceGroups/frontdoor/providers/Microsoft.Network/frontdoors/bink-frontdoor"

Error: flattening `backend_pool`: ID was missing the `healthProbeSettings` element
```

If I look on https://resources.azure.com/ I can see that my Front Door has `healthProbeSettings` with entirely lower case IDs, example:

```
              "healthProbeSettings": {
                "id": "/subscriptions/<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/healthprobesettings/healthz"
              },
              "loadBalancingSettings": {
                "id": "/subscriptions/<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/loadbalancingsettings/standard"
              },
```

So, my hunch is that all I need to do is update the AzureRM API to be case Terraform is expecting to workaround this issue, same as I did for the import command.

So, I switched to Read/Write mode on the Resources site, went into edit mode and manually started updating all IDs to be case sensitive, but only on the specific keys Terraform was complaining about, example:

```
"id": "/subscriptions/<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/healthprobesettings/healthz"
```
became
```
"id": "/subscriptions/<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/healthProbeSettings/healthz"
```

Upon re-running my import, Terraform now moans about a totally different object! HURAHH! Progress!

```
Error: flattening `backend_pool`: ID was missing the `loadBalancingSettings` element
```

So, I ended up fixing case on the following IDs:
* `healthProbeSettings`
* `loadBalancingSettings`
* `frontDoorWebApplicationFirewallPolicies`
* `frontendEndpoints`

Running a Terraform Import between each modification, slowly getting further to success.

Now that front door was fixed, I tried reimporting our WAF policy and was hit with a familiar message:
```
Error: Error parsing Resource ID "/subscriptions/<snipped>/resourcegroups/frontdoor/providers/Microsoft.Network/frontdoorwebapplicationfirewallpolicies/policy": ID was missing the `frontDoorWebApplicationFirewallPolicies` element
```

This was simply a case of fixing my import command to be case-sensitive as with the original import example.

After these imports I was able to run a `terraform plan`, it wanted to apply changes but these were all case related changes so I let it proceed. It did panic the provider but after a few runs of `terraform apply` it got itself unstuck. I was getting messages like:

```
Error: Provider produced inconsistent final plan

When expanding the plan for
module.uksouth-frontdoor.azurerm_frontdoor.frontdoor to include new values
learned so far during apply, provider
"registry.terraform.io/hashicorp/azurerm" produced an invalid new value for
.frontend_endpoint[6].web_application_firewall_policy_link_id: was
cty.StringVal("/subscriptions/<snipped>/resourcegroups/frontdoor/providers/Microsoft.Network/frontDoorWebApplicationFirewallPolicies/policy"),
but now
cty.StringVal("/subscriptions/<snipped>/resourceGroups/frontdoor/providers/Microsoft.Network/frontDoorWebApplicationFirewallPolicies/policy").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

But I guess this makes sense if it's changing objects in this weird example of case-inconsistency.

Anyway, issue fixed for me. I really hope the above helps others and I really hope in the mean time @tombuildsstuff considers merging #8046.

**Comment by GarethOates** [comment]
So @tombuildsstuff rejected the inclusion of the fix in #8046 for his change #8146 which was included in provider version 2.24.0,  but didn't actually fix the problem?

**Comment by tombuildsstuff** [comment]
@cpressland indeed - unfortunately this is a bug on the Azure side which needs fixing.

The HTTP Specification defines that URI's in Requests and Responses must be case sensitive (although servers may opt to parse Request URI's case insensitively, as IIS has done historically). Whilst some Azure API's do support case-insensitively - others require that URI's are sent case sensitively and behave incorrectly (e.g. accepting the data, but failing in subtle ways) if these are in the incorrect casing - ultimately this means that we need these to be case-sensitive in all cases.

In the case of this bug, the issue is the Portal/FrontDoor API's are returning these case insensitively (the root-cause being a misinterpretation of ambiguous wording in the Azure API Specifications) - as such this issue needs a fix on the Microsoft side (in the FrontDoor API) - and as such we have no plans to merge #8046, as previous attempts at doing so caused subtle issues in other API's.

@WodansSon could you reach out to the Portal/Service Teams for an update here, since this has been quite a while since we've had an update from them?

Thanks!

**Comment by GarethOates** [comment]
> and as such we have no plans to merge #8046, as previous attempts at doing so caused subtle issues in other API's.

I understand your desire to have the problem fixed at the root, but so much time has passed since this issue was raised and there are clearly a lot of users who want to be able to manage FrontDoor through terraform.  Can this fix not be merged in as a temporary work around until a more permanent solution is offered from the Microsoft team?  As far as I can see, it's a localized change to a front door specific file. There are so many features since provider 2.23.0 came out that many users will most likely be wanting to take advantage of, but cannot just now due to this bug.

**Comment by cpressland** [comment]
@tombuildsstuff - fair enough, from a technical perspective I completely agree that this is Microsofts burden to fix, I just don't see it actually happening in a sane timeframe. Else, could we mark my previous post as on-topic again? It does include potential workarounds to this issue for users currently blocked by this. I'm happy to edit it to make it more on-topic or even put it somewhere else if you have a good suggestion?

**Comment by tombuildsstuff** [comment]
@GarethOates @cpressland we've tried that historically and it leads to further cases of this, which ultimately ends up breaking other usages in subtle ways.

It's worth calling out that the Azure Networking team get a lot of exceptions, covering:

* Swagger (Azure/azure-rest-api-specs)
  * Where the linting is allowed to fail, due to API's which don't comply with the ARM Spec
* Resource Manager API's
  * Where breaking changes are frequently shipped in existing API versions, when the ARM Spec requires breaking changes go into a new API version
  * Where the ordering of keys in the serialized JSON changes the behaviour of the API (e.g. the API behaves differently when `foo` is serialized before `bar`) - which the JSON spec disallows (which requires that maps/dictionaries are order-insensitive)
* Terraform
  * Where we end up having to work around said breaking changes (which means API version upgrades take longer) and try to identify these serialization issues

Whilst I appreciate it's frustrating to be blocked here - we can't keep layering workarounds on top of workarounds, doing so is exacerbating the problem and leading to the issues we see today (and a bunch of more subtle, and harder to diagnose, issues like the JSON ordering issue mentioned above).

We've reached out to the Portal Team who've committed to fixing this bug on their side (which we've chased them on) - after which we can then work with the Networking Team to fix the root-cause of these bugs.

As Microsoft have committed to fixing these bugs (and for the reasons outline above) - unfortunately we have no plans to introduce a hack for this API bug. In the interim, since this bug is only triggered when editing the resource in the Portal - I believe it should be possible to workaround this using RBAC (or a Management Lock).

From our side, whilst we appreciate this isn't ideal (and is frustrating to be blocked) - we're working with Microsoft to fix this and will post an update as soon as we have one.

Thanks!

**Comment by surlypants** [comment]
once you go through the workaround I described previously, you can upgrade providers.

**Comment by bradcavanagh-unity3d** [comment]
I've just had to upgrade to 2.33.0 (from 2.2!) because of [this issue](https://github.com/Azure/azure-rest-api-specs/issues/11271), and now we're seeing this bug block our deployments. I know that you ultimately have to wait for Microsoft to fix the underlying problem, but could you at least in the meantime mark comments that have workarounds as "on topic" so they're easy to find?

**Comment by GarethOates** [comment]
> 
> 
> > @WodansSon would you be able to get an ETA for a fix from the service team here?
> 
> @tombuildsstuff I will reach out to the Front Door service team and see how quickly we can get a fix in place for this issue, it will most likely include some cross team collaboration with the portal team to roll back their changes to get this issue totally fixed.

Did you ever hear back from the Microsoft Portal team about this issue?  What's the ETA on a fix from their end?

**Comment by Rorschachly** [comment]
I am also wondering if this fix could be applied sooner. Just as @GarethOates thought. Thanks.

**Comment by ajklotz** [comment]
I'm blocked by this issue and I cannot downgrade as some others have mentioned because I require azurerm resources that are not available in 2.23. As of now, managing the Azure Front Door is impossible and I'm surprised that azurerm_frontdoor is in this state

**Comment by tombuildsstuff** [comment]
@WodansSon did you get an update from the Service Team here?

**Comment by WodansSon** [comment]
@tombuildsstuff 

> @WodansSon did you get an update from the Service/Portal Team here?

The last I heard from the service team, Nov. 4th 2020, is that the deployment of the fix is ongoing and completed for several stages but there was an Azure locked down blocking the complete roll out of the fix to all regions. They stated that they would continue to roll out the change to the rest of the regions after the lock down was lifted. I will continue to put pressure on the portal team in an attempt to get this pushed out ASAP.

**Comment by WodansSon** [comment]
@andrstor @tombuildsstuff The fix has be deployed to all regions, so I am going to go ahead and close this issue. Sorry it took so long for the turnaround. 🚀

**Comment by alec-pinson** [comment]
Hi, I think I'm still getting the same issue, is anyone able to confirm this is fixed for them?  
We're in the North Europe region.  

When trying to import I get the following:-  
```
Error: Error parsing Resource ID "/subscriptions/0000000-0000-000-0000-000/resourceGroups/NEUR-RG/providers/Microsoft.Network/frontdoors/my-frontdoor": ID was missing the `frontDoors` element
```
Moved from another state I had already and:-  

```
Error: flattening `frontend_endpoint`: ID was missing the `frontDoorWebApplicationFirewallPolicies` element
```

**Comment by camallen** [comment]
I'm still seeing the previous errors as well on our TF front door resources
```
Error: flattening `backend_pool`: ID was missing the `healthProbeSettings` element
Error: flattening `routing_rules`: flattening `frontend_endpoints`: ID was missing the `frontendEndpoints` element 
```

**Comment by Poil** [comment]
Same here, still have problem

**Comment by tombuildsstuff** [comment]
@WodansSon given that was just the portal, is there a timeline for the API fix too?

**Comment by WodansSon** [comment]
> @WodansSon given that was just the portal, is there a timeline for the API fix too?

The API on the service side was fixed months ago, it was the Portal part that took more time to release. I suspect this maybe an issue with legacy resources that were created before either were fixed. I will need to figure out how to create a Frontdoor service with these issues in order to get a repro so I can best understand how to mitigate the legacy services. For net new Frontdoors, this has indeed been fixed from my testing with and without WAF.

**Comment by WodansSon** [comment]
> Hi, I think I'm still getting the same issue, is anyone able to confirm this is fixed for them?
> We're in the North Europe region.
> 
> When trying to import I get the following:-
> 
> ```
> Error: Error parsing Resource ID "/subscriptions/0000000-0000-000-0000-000/resourceGroups/NEUR-RG/providers/Microsoft.Network/frontdoors/my-frontdoor": ID was missing the `frontDoors` element
> ```
> 
> Moved from another state I had already and:-
> 
> ```
> Error: flattening `frontend_endpoint`: ID was missing the `frontDoorWebApplicationFirewallPolicies` element
> ```

@alec-pinson, the casing in the import command appears to be wrong. Have you tried importing your resource with `frontDoors` instead of `frontdoors`? I suspect, as I stated above, this has to do with existing resources that were created prior to the fixes being released by both teams. I will need to dig deeper into this and see if I can come up with migration path that will correct the current states.

Also of **note:** The provider version should be the latest version as there have been changes made to this resource.

**Comment by WodansSon** [comment]
@camallen @Poil @alec-pinson @tombuildsstuff 

I have done a bit of an experiment and what I have found is the easiest way to fix a legacy resource that is stuck in this in between state is to modify the resource in the Portal and save the changes. Once the changes are saved you can then manage the resource via Terraform. When I was investigating this I disabled one of my routing rules, update, then saved. Once the modification to the service was complete via portal I was then able to successfully manage it again with Terraform. You have to modify the configuration of the Frontdoor directly, adding or updating a Tag will not trigger the code that rewrites the id's with the correct casing.

**NOTE:** The provider version should be the latest version as there have been changes made to this resource to account for various other casing issues.

**Comment by Poil** [comment]
After editing in the portal I still have a problem on `frontDoorWebApplicationFirewallPolicies` that is still `frontdoorWebApplicationFirewallPolicies` in the resource explorer

**Comment by alec-pinson** [comment]
Hi, same for me, I disabled a routing rule, tried terraform and got the error below, then enabled the routing rule and tried again just incase, but same thing.

```
Error: flattening `frontend_endpoint`: ID was missing the `frontDoorWebApplicationFirewallPolicies` element
```

When trying to import (I corrected frontDoors casing, thank you!) 
And same if I try to use a state that already has the frontdoor resource


provider versions

```
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/azuread v0.11.0
+ provider registry.terraform.io/hashicorp/azurerm v2.36.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
```

**Comment by GarethOates** [comment]
Maybe this should be re-opened?

**Comment by naikajah** [comment]
I agree this ticket should be re-opened. I tried updating the provider to the latest 2.36.0 and have the same issue.

`Error: flattening `frontend_endpoint`: ID was missing the `frontDoorWebApplicationFirewallPolicies` element`

**Comment by timja** [comment]
is the remaining issue perhaps just in `frontDoorWebApplicationFirewallPolicies`? 

As the last 3 reports are mentioning that while the initial report was about `healthProbeSettings`

**Comment by GarethOates** [comment]
> is the remaining issue perhaps just in `frontDoorWebApplicationFirewallPolicies`?
> 
> As the last 3 reports are mentioning that while the initial report was about `healthProbeSettings`

You'd like to think MS would have seen the bigger picture and taken the opportunity to fix the casing across all properties where it could be a problem, not just the individual one mentioned, but you never know..

**Comment by kplantus** [comment]
@WodansSon 
I'm throwing my hat in as another who still has an issue with the FD WAF policy.

I even tried to create a brand new AFD.  After it created, I modified a value via the portal to see if the rewrite workaround would work for a new resource.
Still get:
```
Error: flattening `frontend_endpoint`: ID was missing the `frontDoorWebApplicationFirewallPolicies` element
```

**Comment by WodansSon** [comment]
When reporting issues, please keep in mind to remain respectful and professional, I understand that this is frustrating and I am doing the best that I can to correct the issue, however some aspects are out of my control as they are not under my direct control. Since WAF is it's own separate resource I going to assume the workaround will have to be applied to that resources as well(e.g. open WAF in portal, update, save) to correct the casing issues in that resource. Since it is it's own resource I will have to reach out to the service team again to verify that the fix that was applied to that resource as well. I will open a new issue for the WAF specifically since this issue for frontdoor appears to be fixed.

**Comment by WodansSon** [comment]
> I agree this ticket should be re-opened. I tried updating the provider to the latest 2.36.0 and have the same issue.
> 
> `Error: flattening `frontend_endpoint`: ID was missing the `frontDoorWebApplicationFirewallPolicies` element`

I disagree, since the WAF is a different resource I believe it should be a new issue.

**Comment by kplantus** [comment]
Did a little bit more testing and found that on the brand new AFD I could modify a few things via https://resources.azure.com and bring it under TF management using TF v13.5 and azurerm provider v2.36.0.

under `"frontendEndpoints" / "webApplicationFirewallPolicyLink"` I updated the `id` of each FE from all lowercase to `frontDoorWebApplicationFirewallPolicies`.

The next plan failed on:

```
Error: flattening `routing_rules`: flattening `frontend_endpoints`: ID was missing the `frontendEndpoints` element
```

so back in resources under `"routingRules" / "frontendEndpoints"` I updated the `id` of each RR from all lowercase to `frontendEndpoints`.

My next tf plan and apply worked.

**Comment by WodansSon** [comment]
@kplantus Yes, that is exactly what I was suspecting was going on as well. I have don't some digging myself and found this appears to be an issue in the API instead of Portal this time. I am already in contact with the service team to get an ETA for a high pri fix and deployment. However, that is still in negotiations with the team, at the same time I am also in contact with the portal team to ensure that the lower casing of the ID issue isn't also in their UI layer.

**Comment by naikajah** [comment]
Thanks for the update @WodansSon 

I can confirm the steps mentioned by @kplantus does not work with an existing AFD. I also tried deleting the linked WAF policies with the AFD and then running the terraform apply but got the same error. Will wait for the fix.

**Comment by WodansSon** [comment]
> Thanks for the update @WodansSon
> 
> I can confirm the steps mentioned by @kplantus does not work with an existing AFD. I also tried deleting the linked WAF policies with the AFD and then running the terraform apply but got the same error. Will wait for the fix.

Hi @naikajah, what version of the provider are you using? I updated my comment above to state that the provider should be at least v2.24.0.

**Comment by camallen** [comment]
I can confirm the [steps mentioned](https://github.com/terraform-providers/terraform-provider-azurerm/issues/8208#issuecomment-729938968 ) by @kplantus (thank you 👍 ) do work for us on our existing AFD resources. We do not have any WAF configured. 

Tested with v2.36.0 of the resource provider and Terraform v0.13.5

It's not ideal to edit the resources directly in the azure portal and I'm not sure what will happen if we edit the AFD resources in the portal again, I assume we might re-break the AFD resource definitions.

Hopefully this is useful for the Azure portal team and helps someone else get TF working again.

**Comment by WodansSon** [comment]
@camallen Your AFD will be fine now since you corrected the casing that was introduced by the portal lowercase issue. You are safe to edit the Frontdoor via portal as well since they have reverted the changes they made that caused this issue in the first place. 

**UPDATE:** I have just heard back from the portal team and they have confirmed that the lower casing issue also exists in the Frontdoor Web Application Firewall Policies UI layer and they are currently working on a fix for that. I have yet to hear back from the AFD team, but I will keep you posted on the progress of this issue once I receive more information. Thank you. 🚀

**Comment by windsurfer123** [comment]
We have a very similar problem. We had to upgrade recently to 2.24 because of APIM issues and now when we are trying to manage the existing front door resources we get errors. 
Set up: tf either 13.5 or 12.9 azurerm v 2.37
Problem 1: when creating resources terraform sits on the "Still creating ..." for over an hour, at which point it is terminated. All objects are getting created properly. On subsequent runs it want to update the FD, which of course never finishes and has to be terminated. 
Problem 2: when trying to import the Frond Door resource created by terraform in Problem 1, we get Error: Error parsing Resource ID ""v ID was missing the `frontDoors` element

**Comment by brunoscota** [comment]
same happening here. I had to rollback azurerm version to v2.23.0 as a workaround. Now, I am stuck. ;(

**Comment by camallen** [comment]
> @camallen Your AFD will be fine now since you corrected the casing that was introduced by the portal lowercase issue. You are safe to edit the Frontdoor via portal as well since they have reverted the changes they made that caused this issue in the first place.

Sadly, after editing our FD resources in the Azure portal we now have the same old error and broken TF
> Error: flattening `routing_rules`: flattening `frontend_endpoints`: ID was missing the `frontendEndpoints` element`

Looks like the portal is still changing the resource definition from `frontendEndpoints` to non expected case for the Azure provider. 

E.g. in the https://resources.azure.com/ portal this is what I see (with redactions) for one of the broken FD resource routing rules, note the different caching on different routing rules
```
{
  "id": "/subscriptions/.../resourcegroups/.../providers/Microsoft.Network/Frontdoors/.../FrontendEndpoints/my-custom-domain-org-azurefd-net"
},
{
  "id": "/subscriptions/.../resourcegroups/.../providers/Microsoft.Network/frontdoors/.../frontendendpoints/my-custom-domain"
}
```

**Comment by KyMidd** [comment]
Hey @WodansSon and $MSFT internal team, is it possible to provide an ETA for when this collection of resources will be repaired at the API layer? So far we're getting updates after a fix is made but nothing to give us an estimate of how long we'll need to wait for this fix to be implemented. Are we looking at days/weeks/months? 

The healthcare teams I support use these resources for security and the time-frame you provide can help inform if we should switch to an alternate IaC solution or other workarounds. 

Thanks in advance! 
kyler

**Comment by WodansSon** [comment]
@KyMidd Thank you for your question. I am still attempting to get in contact with the Front Door service team about this and have begun to escalate the issue internally to force some action about getting this fixed. Unfortunately I am not able to provide an ETA at this time, but as soon as I have any new news I will promptly update the status here.

**Comment by subesokun** [comment]
Somehow I'm now in a kind of dead lock. I need to downgrade to `2.23.0` because of this bug but when I do so I'm running into another frontdoor related TF bug (https://github.com/terraform-providers/terraform-provider-azurerm/issues/8036) which was solved in `2.24.0`. Very frustrating.

**Comment by WodansSon** [comment]
@subesokun I am sorry to hear that, but please be patient, @tombuildsstuff and myself have been working on this issue and I believe we have a solution in the provider which should fix 99.999% if the issues that are currently being hit... We are still testing the solution, but so far it all looks good... again, I am sorry for this pain, but we are doing all we can to correct this issue.

**Comment by tombuildsstuff** [comment]
Ultimately this'll be fixed via #9750 - which ignores the casing returned from the Azure API and rewrites this to be consistent on Terraform's side, whilst there's downsides to that approach (and the Service Team ultimately need to fix these bugs in the API..) this should workaround this series of API bugs for the moment.

**Comment by KyMidd** [comment]
@WodansSon and @tombuildsstuff : Sincere thank you from myself and on behalf of the community here for working on a workaround for this issue. I have several teams affected, and despite us pushing hard on Microsoft's internal enterprise support, we haven't made any progress at all. I think we're all aware here that the real root cause of this issue is standard-breaking API implementation on Azure side, and from your user-base I want to say: Thank you.

**Comment by ajklotz** [comment]
@WodansSon @tombuildsstuff Is there a ticket that we can go and add our voice of support? Anything we can do to help MS understand there is attention from all of us regarding it?

**Comment by WodansSon** [comment]
@ajklotz I don't believe there is a public facing ticket or support request for this issue. That said, I have included a link to this issue in all internal communications for this problem so they are very much aware of the contention the changes in the API have caused. Thank you for asking!

**Comment by camallen** [comment]
Thanks @tombuildsstuff & @WodansSon for fixing this 🎩 and thanks to all the folks that contributed to the issue 👍.

At 9th Dec 2020 10:25 UTC it appears that this fix is unreleased https://github.com/terraform-providers/terraform-provider-azurerm/blob/2df9c3193a43380e16f0000f3366b221b31d6c74/CHANGELOG.md#2400-unreleased 

Is there a timeline for this fix being officially released? I'm very keen (and I assume so are a lot of folks) to test these fixes and get TF integrations working again. 

FWIW I can't see how to use this unreleased provider version via https://www.terraform.io/docs/configuration/provider-requirements.html#version-constraints - maybe I'm missing something....

Any ideas or is it wait till the official release?

**Comment by katbyte** [comment]
@cam

**Comment by KyMidd** [comment]
I'm sorry to report that I compiled azurerm provider as of this morning and we are seeing the same type of issues on a refresh where terraform's unable to handle. 
```
K817684-MacOS:ue1 k817684$ terraform refresh
(removed)
Error: expected "custom_rule.0.match_condition.0.selector" to not be an empty string, got 
  on main.tf line 27, in resource "azurerm_frontdoor_firewall_policy" "frontdoor":
  27: resource "azurerm_frontdoor_firewall_policy" "frontdoor" {
Error: expected "custom_rule.1.match_condition.0.selector" to not be an empty string, got 
  on main.tf line 27, in resource "azurerm_frontdoor_firewall_policy" "frontdoor":
  27: resource "azurerm_frontdoor_firewall_policy" "frontdoor" {
```
The FrontDoor and FrontDoor WAF rules were created from terraform using the new provider. Note the differing capitalization of "frontDoor" vs "frontdoor". I redacted the account, otherwise unchanged. 
```
/subscriptions/./resourceGroups/TestingKyler/providers/Microsoft.Network/frontDoorWebApplicationFirewallPolicies/KylerTestingWafPolicy
```
And from the portal: 
```
/subscriptions/./resourceGroups/TestingKyler/providers/Microsoft.Network/frontdoorWebApplicationFirewallPolicies/KylerTestingWafPolicy
```
If possible, I'd like to see this issue reopened. Thanks team.

**Comment by WodansSon** [comment]
@KyMidd Sorry that you are hitting that issue, however the casing should not matter anymore, I believe you are hitting a validation rule. Do you have a repro for this? If I can get a clear repro for this issue I might be able to get a fix in before the next release. Thank you.

**Comment by KyMidd** [comment]
Hey @WodansSon , Oh, I assumed Terraform was modifying the casing on import. 

As for repro, definitely. 
1. Using this config, simple WAF + FrontDoor: https://github.com/KyMidd/azurerm-frontdoor-testing-repro
2. Compile from master azurerm and store properly
3. Authenticate to Azure via az login
4. terraform init (should find local azurerm)
5. terraform apply <-- Resources built properly
6. terraform refresh <-- Terraform fails with error message

Let me know if I can help further by testing a PR. Thank you!

**Comment by KyMidd** [comment]
@WodansSon : Update, I have destroyed and recreated this same config several times now to make sure the same error is generated, and I'm unable to replicate it. I'm not sure what happened there. I am now seeing FrontDoor reliably managed by terraform!

**Comment by WodansSon** [comment]
@KyMidd, I am unable to repro the behavior you are reporting above. Awesome, that would explain my results as well! Awesome! 🚀 

I followed your steps with the same config file and when I execute step 6 I get:

**Refresh**
```
azurerm_resource_group.testing_kyler: Refreshing state... [id=/subscriptions/{subscription}/resourceGroups/XXXXXX-frontDoor-Repro]
azurerm_frontdoor_firewall_policy.frontdoor: Refreshing state... [id=/subscriptions/{subscription}/resourceGroups/XXXXXX-frontDoor-Repro/providers/Microsoft.Network/frontDoorWebApplicationFirewallPolicies/KylerTestingWafPolicy]
azurerm_frontdoor.example: Refreshing state... [id=/subscriptions/{subscription}/resourceGroups/XXXXXX-frontDoor-Repro/providers/Microsoft.Network/frontDoors/XXXXXX-testing-frontdoor]
```

**Plan**
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

azurerm_resource_group.testing_kyler: Refreshing state... [id=/subscriptions/{subscription}/resourceGroups/XXXXXX-frontDoor-Repro]
azurerm_frontdoor_firewall_policy.frontdoor: Refreshing state... [id=/subscriptions/{subscription}/resourceGroups/XXXXXX-frontDoor-Repro/providers/Microsoft.Network/frontDoorWebApplicationFirewallPolicies/KylerTestingWafPolicy]
azurerm_frontdoor.example: Refreshing state... [id=/subscriptions/{subscription}/resourceGroups/XXXXXX-frontDoor-Repro/providers/Microsoft.Network/frontDoors/XXXXXX-testing-frontdoor]

------------------------------------------------------------------------

No changes. Infrastructure is up-to-date.

This means that Terraform did not detect any differences between your
configuration and real physical resources that exist. As a result, no
actions need to be performed.
```

**Comment by jacksondaw** [comment]
This issue was still occurring for me when I added a second frontend.  The first existing frontend was throwing the following error. I tried upgrading to 2.40.0 and downgrading to 2.23.0, both had the same error:

Error: updating Custom HTTPS configuration for Frontend Endpoint "****" (Front Door "" / Resource Group "FrontDoor"): unable to enable/update Custom Domain HTTPS for Frontend Endpoint "***" (Resource Group "FrontDoor"): enabling Custom Domain HTTPS for Frontend Endpoint: frontdoor.FrontendEndpointsClient#EnableHTTPS: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="That action isn’t allowed in this profile."

I ended up pulling the https configuration out into the [frontdoor_custom_https_configuration ](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/frontdoor_custom_https_configuration) and the issue was resolved.
